apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClient.cs

```diff
@@ -482,12 +631,16 @@ private void ShutdownIfCompleted()
                 if (_hardStopCts.IsCancellationRequested)
                 {
                     // Cancel any remaining batches. Only relevant if hard-stopped.
+                    List<Batch> batchesToCancel;
                     lock (_lock)
                     {
-                        foreach (var batch in _batchesReady)
-                        {
-                            batch.BatchCompletion.SetCanceled();
-                        }
+                        batchesToCancel = _batchesReady.Select(x => x.Batch)
+                            .Concat(_keyedState.Values.SelectMany(state => state.Batches))
+                            .ToList();
+                    }
+                    foreach (var batch in batchesToCancel)
+                    {
+                        batch.BatchCompletion.SetCanceled();
```

Comment:
I assume it's deliberate that this is outside the lock, when it was inside before?